keps/sig-storage/5040-remove-gitrepo-driver/README.md

```diff
@@ -0,0 +1,889 @@
+know that this has succeeded?
+-->
+
+- Remove in-tree gitRepo volume support.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Kubernetes drops support for gitRepo volume types by removing the in-tree driver
+that actually mounts the volume in the pod. In this case we will still keep the
+gitRepo volume in the API. This removal is similar to removal of support for
+[glusterfs](https://kubernetes.io/docs/concepts/storage/volumes/#glusterfs)
+which was deprecated in v1.25 and removed in v1.26.
+
+Sig-storage was actually very close to doing this and had even added a notice of
+removal in the release notes for v1.32. But due to concerns raised by sig-arch
+they rolled back the notice. sig-storage provided [this rationale](https://github.com/kubernetes/kubernetes/issues/125983#issuecomment-2522201283) for considering removal.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The biggest risk is in breaking users. Despite being deprecated for years, there
+may be some users who are still using it and their workloads would be broken 
+when their clusters are upgraded. We plan to mitigate the risk by allowing
+users to opt-in to re-enabling the driver while they fix the workloads.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We will add a new feature-gate to kubelet called `GitRepoVolumeDriver`. This
+feature-gate will be defaulted to `false`, which will disable the 
+[gitRepo driver][gitRepo driver](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/git_repo/git_repo.go).
+If a workload with gitRepo volume is encountered, kubelet will fail to run such a
+pod with an appropriate error message guiding users to either use an alternative
+or setting the feature-gate `GitRepoVolumeDriver` to `true`.
+
+### Timeline:
+v1.33 (04/2025): 
+- We announce the removal in kubernetes release notes, documentation and via
+kubernetes release blog post.
+- We add the `GitRepoVolumeDriver` feature-gate to kubelet and
+default it to `false`.
+
+v1.34 (08/2025):
+- No change.
+- We will announce removal in kubernetes release notes, documentation and via
+kubernetes release blog post.
+
+v1.35 (11/2025):
+- No change.
+- We will announce removal in kubernetes release notes, documentation and via
+kubernetes release blog post.
+
+v1.36 (04/2026):
```

Comment:
For compat versions, we need 3 versions where the behavior can be made the same wrt toggle.  Removing the gate means that it can not be toggled off.  This means we need to lock the gate, wait 3 releases, AND THEN we can remove it.

So if we have the gate enableable in 33, 34, 35 and then lock it to off in 36, we need to wait until 39 to actually remove it.

If we lock it in 34, we can remove in 37, but Jordan suggested leaving it toggleable for more than 1 release.  So... do we leave it toggleable for 3 releases or 2?

@liggitt ?  It's probably fine to leave it for 3, meaning 39 is the removal.

Reply:
I think the only bits that are urgent are 1) disabling by default with an escape hatch and 2) starting the process to drive feedback to inform a decision on locking / completely removing.

Once we've disabled by default, I don't especially care if we leave the inert plugin there and opt-in ~forever.

I don't know how long to leave between 1 and 2... enough time that impacted people reach the minor that disables and requires opt-in. I'd guess 2-3 releases.

Reply:
I agree with @liggitt that the urgent thing is disabling by default with an escape hatch and getting feedback from users before we lock. Updated the timelines to lock in 1.36 and remove in 1.39